tools/kolla-images.py

```diff
@@ -0,0 +1,365 @@
+    """Return a list of containers that use the specified image."""
+    default_container = image.replace('-', '_')
+    return IMAGE_TO_CONTAINERS_EXCEPTIONS.get(image, [default_container])
+
+
+def get_parent_tag_name(kolla_image_tags: KollaImageTags, base_distro: Optional[str], container: str) -> str:
+    """Return the parent tag variable for a container in the tag variable hierarchy."""
+
+    if container in CONTAINER_TO_PREFIX_VAR_EXCEPTIONS:
+        parent = CONTAINER_TO_PREFIX_VAR_EXCEPTIONS[container]
+        if parent in kolla_image_tags:
+            return parent
+
+    def tag_key(tag):
+        """Return a sort key to order the tags."""
+        if tag == "openstack":
+            # This is the default tag.
+            return 0
+        elif tag != container and container.startswith(tag) and (base_distro is None or base_distro in kolla_image_tags[tag]):
+            # Prefix match - sort by the longest match.
+            return -len(tag)
+        else:
+            # No match.
+            return 1
+
+    return sorted(kolla_image_tags.keys(), key=tag_key)[0]
+
+
+def get_parent_tag(kolla_image_tags: KollaImageTags, base_distro: str, container: str) -> str:
+    """Return the tag used by the parent in the hierarchy."""
+    parent_tag_name = get_parent_tag_name(kolla_image_tags, base_distro, container)
+    return kolla_image_tags[parent_tag_name][base_distro]
+
+
+def get_tag(kolla_image_tags: KollaImageTags, base_distro: str, container: str) -> str:
+    """Return the tag for a container."""
+    container_tag = kolla_image_tags.get(container, {}).get(base_distro)
+    if container_tag:
+        return container_tag
+
+    return get_parent_tag(kolla_image_tags, base_distro, container)
+
+
+def get_tags(base_distros: List[str], kolla_image_tags: KollaImageTags) -> Dict[str, List[str]]:
+    """Return a list of tags used for each image."""
+    images = read_images("etc/kayobe/pulp.yml")
+    unbuildable_images = read_unbuildable_images("etc/kayobe/pulp.yml")
+    image_tags: Dict[str, List[str]] = {}
+    for base_distro in base_distros:
+        for image in images:
+            if image not in unbuildable_images[base_distro]:
+                for container in get_containers(image):
+                    tag = get_tag(kolla_image_tags, base_distro, container)
+                    tags = image_tags.setdefault(image, [])
+                    if tag not in tags:
+                        tags.append(tag)
+    return image_tags
+
+
+def get_openstack_release() -> str:
+    """Return the OpenStack release."""
+    with open(get_abs_path(".gitreview"), "r") as f:
+        gitreview = f.readlines()
+    for line in gitreview:
+        if "=" not in line:
+            continue
+        key, value = line.split("=")
+        if key.strip().strip() == "defaultbranch":
+            value = value.strip().rstrip()
+            prefix = "stable/"
+            assert value.startswith(prefix)
+            return value[len(prefix):]
```

Comment:
Maybe add support for the `unmaintained/` prefix already?

Reply:
I'll take a look at this.

Reply:
It may be required only if:

1) upstream kayobe-config changes `stable/yoga` to `unmaintained/yoga`. Would Yoga be marked eol instead?
2) we merge upstream kayobe-config into stackhpc-kayobe-config (we don't routinely do this at the moment)